Add admin interface via ActiveAdmin

lumpidu opened the pull request:
Provide standard configuration for ActiveAdmin and make available the resources conversations, messages.

So far very basic and raw, but already usable for cleaning up and debugging. Authentication & authorization is done
via the [Devise](https://github.com/heartcombo/devise) gem.

For development, the admin user/password is put as a fixed string into [seed.rb](db/seed.rb), but for production is derived from the environment variables `ADMIN_USER`, `ADMIN_PASSWORD` at `bin/rails db:seed` time.

Moreover change the database type of db model `message#reply` from `string` to `jsonb` and convert existing data of that column to `JSON` inside the db migration. In the same context, fix generation of the `message#reply` `JSON` data inside `conversation#create`.

- [x] user authentication with Devise
- [x] admin credentials via environment variables
- [x] augment DB messages model with dialog system metrics
- [ ] show conversation content as user/bot dialog
- [x] show graph with metrics on dashboard
- [x] remove API endpoints that can be easier managed via the admin dashboard

![grafik](https://github.com/SDiFI/masdif/assets/734966/f210211b-7161-4720-9e7c-b109e8ebcdc4)

lumpidu committed:
Add basic admin interface via ActiveAdmin

Provide standard configuration for ActiveAdmin and make available
the resources conversations, messages.

So far very basic and raw, but already usable for cleaning up and
debugging.

Moreover change the database type of message#reply from string to jsonb
and convert existing data of that column to jsonb inside the db migration.
In the same context, fix generation of the message reply JSON data inside
conversation#create

Signed-off-by: Daniel Schnell <dschnell@grammatek.com>

lumpidu commented on db/seeds.rb:
```diff
@@ -5,3 +5,18 @@
+  admin_user = ENV['ADMIN_USER']
+  admin_password = ENV['ADMIN_PASSWORD']
+  if admin_user && admin_password
+    User.create!(email: admin_user, password: admin_password, password_confirmation: admin_password)
```
Password is provided via environment variable, therefore not applicable.

lumpidu commented on db/seeds.rb:
```diff
@@ -5,3 +5,18 @@
+  admin_user = ENV['ADMIN_USER']
+  admin_password = ENV['ADMIN_PASSWORD']
+  if admin_user && admin_password
+    User.create!(email: admin_user, password: admin_password, password_confirmation: admin_password)
```
Password is provided via environment variable, therefore not applicable.

lumpidu committed:
Every API controller now inherits from ActionController::API.

This removes some middle-ware layers that are normally necessary. E.g.
the CSRF token check is omitted. Therefore, we need to come up soon
with an authentication approach for the controllers.


Signed-off-by: Daniel Schnell <dschnell@grammatek.com>

lumpidu committed:
Process custom actions from the dialog system.

If a key value 'custom' is present inside the responses array, the value
of that key is an action reply containing JSON. Save the action reply to
the message and remove the custom element from the reply attribute.

Also do this for existing db columns and add an appropriate migration.

Signed-off-by: Daniel Schnell <dschnell@grammatek.com>

lumpidu committed:
Additional properties for message model: NLU, Tracker events, time measuremnts

Query Rasa tracker, filter unnecessary/redundant infos and save NLU,
events.

Signed-off-by: Daniel Schnell <dschnell@grammatek.com>

lumpidu committed:
Configure/style Admin interface: Dashboard, Conversations, Messages

- add graphs to Dashboard for intents, entities, actions and overall
  activity
- reduce shown fields of Conversations index and make time range
  selectable
- reduce shown fields of Messages of a conversation#show action
- remove comments from AA

AA settings persistence:

- store current filter settings into session storage so that these survive
  navigating away to a different view
- also store selected scopes to session storage and extract common time
  scopes to models/concerns/time_scopes.rb for usage inside messages.rb /
  conversations.rb
- pagination is als stored inside session storage

Signed-off-by: Daniel Schnell <dschnell@grammatek.com>

lumpidu committed:
Remove admin-routes from public access

- restrict the public converation endpoints to just the ones accessible
  from the webchat widget

The admin interface uses the database direclty, so it doesn't need to access
conversation routes.

The only possibility of manipulating conversations from within the admin
interface is (bulk-) deleting conversations including associated messages
and attachments.

Also, remove functional tests for deleted endpoints

Signed-off-by: Daniel Schnell <dschnell@grammatek.com>

lumpidu committed:
Dashboard: add statistics and add positive/negative values to Activity

app/models/message.rb

- add self.feedback_counts() that counts the number of messages that have
  been voted on via user feedback and returns the count as a Hash
- add self.feedback_date_series() that returns a time series for the
  feedbacks compatible with groupdate gem and therefore ApexCharts
- self.stats_counts(): add values for asr_count, tts_count

db/schema.rb

- add indexes for feedback and text and an approprate migration

app/admin/dashboard.rb:

- add Stats panel with overall counts for messages in selected time perid,
  asr + tts stats as well as feedback counts
- add feedback graphs to panel Activity and refactor out the charts into
  own partial app/views/admin/dashboard/_chart.html.erb

Signed-off-by: Daniel Schnell <dschnell@grammatek.com>

lumpidu committed:
Fix DB message::reply schema handling

The reply column was wrongly written as String value instead of a real
jsonb value, because before saving it to the db, the hash value has been
converted to a string via #to_json.

- add migration to convert all existing DB records to correct jsonb or nil
- adapt code that reads/writes the column to correctly handle jsonb
- also use a default jsonb value of {} for nil values and convert all
  existing nil values
- add asr_generated: false, tts: false to empty_response_message() metadata

Signed-off-by: Daniel Schnell <dschnell@grammatek.com>

lumpidu committed:
Admininistration screen for Message resource adapted

- rearrange filters, rename columns
- add column highlighting for user feedback 'positive', 'negative'
- add filters for user text, bot answer, intent, ASR used and the
  possibility to filter for user/internal messages - messages that start
  with '/'

Signed-off-by: Daniel Schnell <dschnell@grammatek.com>

lumpidu committed:
Fix migration security problem

lumpidu committed:
Migration: add default value {} to all jsonb columns and make it non-null

lumpidu committed:
app/model/message.rb: fix Dashbaord Activity rendering for missing feedback

- feedback_date_series() is adapted to always include hash values for
  all required feedback values. Therefore, each found date will have at
  least a data point for each feedback value of value 0
- use scope .exclude_internal instead of an SQL clause

Signed-off-by: Daniel Schnell <dschnell@grammatek.com>

lumpidu committed:
Update Admin conversation screen

- add filters similar to the messages screen
- show a Feeback column with a small pie chart that shows the distribution
  ration of positive to negative feedback values
- Replace the id column with an excerpt of User messages
- Column "Update At" shows the timestamp of the last message of a
  conversation. This is made possible by adding an :after_save hook to
  update a new field  'last_message_updated_at' of the conversation model
- add an appropriate migration for the new field and update all existing
  conversations as well.

Signed-off-by: Daniel Schnell <dschnell@grammatek.com>

bnika commented on README.md:
```diff
@@ -24,21 +24,19 @@ build of Github, if a new tag is pushed to this repository.
-- `DELETE /conversations/:id` - Deletes conversation with given id, TODO: needs to be protected
+- `GET /health` - Returns the health status of the API
 - `GET /info` - Returns information about the bot. This is typically the first request sent by the client and is
    used to initialize the chat widget with e.g. the motd and supported languages.
```
write out "message-of-the-day"?

lumpidu replied:
done

bnika commented on config/initializers/aa_filters_persistence.rb:
```diff
@@ -0,0 +1,27 @@
+# see https://pawelurbanek.com/active-admin-tips-optimizations
```
what does the "aa" prefix mean?

lumpidu replied:
active_admin :)

lumpidu committed:
Added Masdif version info

Added version info for Masdif and show in Admin interface as well as
provide a GET /version endpoint to the RESTful API. The version is
derived either from the Git repository, if available or alternatively by
reading environment variables, if e.g. building a Docker image.

- add custom ActiveAdmin header implementation for integrating the version
  info
- add app/helpoers/version_helper.rb for handling either Git-based or
  environment variable based version info
- use Rugged Gem for directly handling Git operaations within the Rails
  process instead of forking via system call
- add functional test for GET /version
- docker/masdif/Dockerfile: add cmake dependency

Signed-off-by: Daniel Schnell <dschnell@grammatek.com>

lumpidu committed:
Update README.md:

lumpidu committed:
Add initial Admin interface credentials via Rails credentials

Add the possibility to use Rails credentials additionally to using
environment variables.

The precedence is Rails credentials => Env. variables.

Update documentation in README.md for this feature.

Signed-off-by: Daniel Schnell <dschnell@grammatek.com>

bnika commented on test/controllers/conversations_controller_test.rb:
```diff
@@ -117,13 +94,6 @@ def check_meta_data(json_response, kwargs)
-    end
-    assert_response :success
-  end
-
```
why the deleted tests?

lumpidu replied:
I have removed public endpoints, because of the admin interface that is a much better place to review, delete conversations. Previously, everybody in the world could do that :)
